File: frontend/src/types/product.ts
import type { SupplierSummary } from "@/types/supplier";

export interface Product {
  id: string;
  name: string;
  sku: string | null;
  category: string | null;
  description: string | null;
  current_quantity: number;
  min_stock: number;
  cost_price: string;
  sale_price: string;
  supplier: SupplierSummary | null;
  is_active: boolean;
  created_at: string;
  is_low_stock: boolean;
}

export interface ProductPayload {
  name: string;
  sku?: string | null;
  category?: string | null;
  description?: string | null;
  min_stock?: number;
  cost_price?: string;
  sale_price: string;
  supplier_id?: string | null;
  initial_quantity?: number;
}

export type MovementType = "ENTRY" | "EXIT" | "ADJUSTMENT" | "LOSS" | "SALE" | "RETURN";
export type ManualMovementType = "ENTRY" | "EXIT" | "LOSS" | "SALE" | "RETURN";

export interface InventoryMovement {
  id: string;
  product_id: string;
  type: MovementType;
  quantity_change: number;
  quantity_before: number;
  quantity_after: number;
  reason: string;
  created_by_user_id: string | null;
  created_at: string;
}

File: backend/app/schemas/product.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, computed_field

from app.schemas.supplier import SupplierSummary


class ProductCreate(BaseModel):
    name: str = Field(min_length=2, max_length=150)
    sku: str | None = Field(default=None, max_length=50)
    category: str | None = Field(default=None, max_length=100)
    description: str | None = None
    min_stock: int = Field(default=0, ge=0)
    cost_price: Decimal = Field(default=Decimal(0), ge=0)
    sale_price: Decimal = Field(ge=0)
    supplier_id: uuid.UUID | None = None
    initial_quantity: int = Field(default=0, ge=0)


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=150)
    sku: str | None = Field(default=None, max_length=50)
    category: str | None = Field(default=None, max_length=100)
    description: str | None = None
    min_stock: int | None = Field(default=None, ge=0)
    cost_price: Decimal | None = Field(default=None, ge=0)
    sale_price: Decimal | None = Field(default=None, ge=0)
    supplier_id: uuid.UUID | None = None
    is_active: bool | None = None


class ProductRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    sku: str | None
    category: str | None
    description: str | None
    current_quantity: int
    min_stock: int
    cost_price: Decimal
    sale_price: Decimal
    supplier: SupplierSummary | None
    is_active: bool
    created_at: datetime

    @computed_field
    @property
    def is_low_stock(self) -> bool:
        return self.current_quantity <= self.min_stock
